contrib/storage-ipfs/src/main/java/org/apache/drill/exec/store/ipfs/IPFSHelper.java

```diff
@@ -0,0 +1,326 @@
+        timeouts.get(IPFSTimeOut.FIND_PEER_INFO)
+    );
+    if (!result.containsKey("Path")) {
+      return null;
+    }
+
+    // the path returned is of form /ipfs/Qma...
+    String hashString = result.get("Path").split("/")[2];
+    return Multihash.fromBase58(hashString);
+  }
+
+  @FunctionalInterface
+  public interface ThrowingFunction<T, R, E extends Exception>{
+    R apply(final T in) throws E;
+  }
+
+  @FunctionalInterface
+  public interface ThrowingSupplier<R, E extends Exception> {
+    R get() throws E;
+  }
+
+  /**
+   * Execute a time-critical operation op within time timeout. Causes the query to fail completely
+   * if the operation times out.
+   * @param op a Function that represents the operation to perform
+   * @param in the parameter for op
+   * @param timeout consider the execution has timed out after this amount of time in seconds
+   * @param <T> Input type
+   * @param <R> Return type
+   * @param <E> Type of checked exception op throws
+   * @return R the result of the operation
+   * @throws E when the function throws an E
+   */
+  public <T, R, E extends Exception> R timedFailure(ThrowingFunction<T, R, E> op, T in, int timeout) throws E {
+    Callable<R> task = () -> op.apply(in);
+    return timedFailure(task, timeout, TimeUnit.SECONDS);
+  }
+
+  public <R, E extends Exception> R timedFailure(ThrowingSupplier<R, E> op, int timeout) throws E {
+    Callable<R> task = op::get;
+    return timedFailure(task, timeout, TimeUnit.SECONDS);
+  }
+
+  private <R, E extends Exception> R timedFailure(Callable<R> task, int timeout, TimeUnit timeUnit) throws E {
+    Future<R> res = executorService.submit(task);
+    try {
+      return res.get(timeout, timeUnit);
+    } catch (ExecutionException e) {
+      throw (E) e.getCause();
+    } catch (TimeoutException e) {
+      throw UserException.executionError(e).message("IPFS operation timed out").build(logger);
+    } catch (CancellationException | InterruptedException e) {
+      throw UserException.executionError(e).build(logger);
+    }
+  }
+
+  /*
+   * DRILL-7753: implement a more advanced algorithm that picks optimal addresses. Maybe check reachability, latency
+   * and bandwidth?
+   */
+  /**
+   * Choose a peer's network address from its advertised Multiaddresses.
+   * Prefer globally routable address over local addresses.
+   * @param peerAddrs Multiaddresses obtained from IPFS.DHT.findprovs
+   * @return network address
+   */
+  public static Optional<String> pickPeerHost(List<MultiAddress> peerAddrs) {
+    String localAddr = null;
+    for (MultiAddress addr : peerAddrs) {
+      String host = addr.getHost();
+      try {
+        InetAddress inetAddress = InetAddress.getByName(host);
+        if (inetAddress.isSiteLocalAddress() || inetAddress.isLinkLocalAddress()) {
+          localAddr = host;
+        } else {
+          return Optional.of(host);
+        }
+      } catch (UnknownHostException ignored) {
```

Comment:
If this is ignored, can/should we log this event?

Reply:
No. The peer is likely to have other addresses that are IP addresses, which can never cause `UnknownHostException`, so it's safe to ignore it. In the rare case where a peer has no IP addresses but an invalid hostname, the call to `pickPeerHost` will return empty. The caller should be responsible to handle a peer that has no addresses. Like what is done here: https://github.com/bdchain/drill/blob/39bab375aefe1b49af9ce358525ebd7c03543231/contrib/storage-ipfs/src/main/java/org/apache/drill/exec/store/ipfs/IPFSGroupScan.java#L444-L446